### ISSUE TITLE
Unbounded, eager enumerations cause Sorbet to loop at runtime

### ISSUE THREAD

**Issue by jafc-stripe** [sorbet/sorbet#1792]
#### Input

https://sorbet.run/#%23%20typed%3A%20true%0A%0A%23%20100%25%20ok%0AT.let(%5B42%5D.cycle.lazy%2C%20T%3A%3AEnumerable%5BInteger%5D)%0A%0A%23%20Freezes%20at%20runtime%0AT.let(%5B42%5D.cycle%2C%20T%3A%3AEnumerable%5BInteger%5D)

```ruby
# 100% ok
T.let([42].cycle.lazy, T::Enumerable[Integer])

# Freezes at runtime
T.let([42].cycle, T::Enumerable[Integer])
```

#### Observed output

```
No errors! Great job.
```

In pry, however:

```
[dev] main:0> T.let([42].cycle.lazy, T::Enumerator[Integer])
=> #<Enumerator::Lazy: ...>
[dev] main:0> T.let([42].cycle, T::Enumerator[Integer])
```

This hangs indefinitely.

#### Expected behavior

Adding `.lazy` to an enumerator shouldn't be required to prevent Sorbet from looping during runtime-- both should terminate successfully.

**Comment by jafc-stripe** [comment]
Interestingly Sorbet _does_ seem to be erasing type parameters-- the same behavior persists even with type mismatches on the inner element:

```
[dev] main:0> T.let([:foo].cycle.lazy, T::Enumerator[Integer])
=> #<Enumerator::Lazy: ...>
[dev] main:0> T.let([:foo].cycle, T::Enumerator[Integer])
# hangs forever
```

Not sure what's causing Sorbet to force the enumerator here.

**Comment by elliottt** [comment]
Thanks!

**Comment by jez** [comment]
Fixed in #1811
